Packages/com.chisel.graph/Chisel/Editor/ChiselGraphInstanceEditor.cs

```diff
@@ -0,0 +1,144 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEditor.Callbacks;
+using UnityEngine;
+using XNodeEditor;
+
+namespace Chisel.Nodes
+{
+    [CustomEditor(typeof(ChiselGraphInstance))]
+    public class ChiselGraphInstanceEditor : Editor
+    {
+        public override void OnInspectorGUI()
+        {
+            var instance = target as ChiselGraphInstance;
+            instance.TriggerOnValidate();
+
+            DrawDefaultInspector();
+
+            if (GUILayout.Button("Edit", GUI.skin.GetStyle("button")))
+            {
+                instance.graph.instance = instance;
+
+                NodeEditorWindow.Open(instance.graph);
+            }
+
+            if (GUILayout.Button("UpdateCSG", GUI.skin.GetStyle("button")))
+            {
+                instance.UpdateCSG();
+            }
+
+            if (GUILayout.Button("Rebuild", GUI.skin.GetStyle("button")))
+            {
+                instance.Rebuild();
+            }
+
+
+            EditorGUI.BeginChangeCheck();
+
+            ChiselGraphPropertyEditor.OnGUI(instance);
+
+            if (EditorGUI.EndChangeCheck())
+            {
+                instance.UpdateProperties();
+                instance.IsDirty = true;
+                instance.UpdateCSG();
+            }
+        }
+    }
+
+    public class ChiselGraphPropertyEditor
+    {
+        static Dictionary<Type, ChiselGraphPropertyDrawer> s_ParameterDrawers;
+
+        static ChiselGraphPropertyEditor()
+        {
+            s_ParameterDrawers = new Dictionary<Type, ChiselGraphPropertyDrawer>();
+            ReloadDecoratorTypes();
+        }
+
+        [DidReloadScripts]
+        static void OnEditorReload()
+        {
+            ReloadDecoratorTypes();
+        }
+
+        static void ReloadDecoratorTypes()
+        {
+            s_ParameterDrawers.Clear();
+
+            // Look for all the valid parameter drawers
+            var types = GetAllTypesDerivedFrom<ChiselGraphPropertyDrawer>()
+                .Where(
+                    t => t.IsDefined(typeof(GraphPropertyDrawerAttribute), false)
+                    && !t.IsAbstract
+                    );
+
+            // Store them
+            foreach (var type in types)
+            {
+                var attr = (GraphPropertyDrawerAttribute)type.GetCustomAttributes(typeof(GraphPropertyDrawerAttribute), false)[0];
+                var decorator = (ChiselGraphPropertyDrawer)Activator.CreateInstance(type);
+                s_ParameterDrawers.Add(attr.propertyType, decorator);
+            }
+        }
+
+        public static void OnGUI(ChiselGraphInstance instance)
+        {
+            if (instance.properties != null)
+                foreach (var property in instance.properties)
+                    using (new EditorGUILayout.HorizontalScope())
+                    {
+                        DrawOverrideCheckbox(property);
```

Comment:
It would be nice that you could set up within the graph which parameters to expose to the user, instead of exposing everything and using checkboxes to choose what to modify

Reply:
When the user add a property node, then it's supposed to be exposed. I will add a constant node for value inside graph only. If we are following unity's shader graph, we are supposed to have a blackboard, where user have a centralized location to set up all the exposed properties.

Reply:
ok, fair enough